### PR TITLE
Add support for aes256-ctr

### PR DESCRIPTION
**Pull request by mavimo**
Hi @pquerna, I do some improvement to the current implementation:

 - Added support for AES256-CTR that is used in OpenSSH implementation
 - Refactor test to use table driven test structure
 - Update to support go module
 - Update travis to run without sudo needed and some CI improvement
 - Move test in a separate package

Let me know if there are something that need to be changed in order to make it mergeable.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module github.com/ScaleFT/sshkeys
+
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/dchest/bcrypt_pbkdf v0.0.0-20150205184540-83f37f9c154a
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/stretchr/testify v1.2.2
+	golang.org/x/crypto v0.0.0-20181106171534-e4dc69e5b2fd
```
Comment (pquerna) [inline go.mod]: i'd like to be able to keep these dependencies free-standing -- i'm not super familiar with go.mod patterns yet (we're still mostly not yet using it).  ideally I won't want this library holding up min versions with things like x/crypto

File: testdata/keys.go [hunk 2]
```diff
@@ -1,11 +1,12 @@
-// Copyright 2014 The Go Authors. All rights reserved.
```
Comment (pquerna) [inline testdata/keys.go]: we shouldn't delete this -- this file is derived from a file in x/crypto/ssh

File: marshal_test.go [hunk 3]
```diff
@@ -1,75 +1,92 @@
...
 
-func TestMarshal(t *testing.T) {
+func TestMarshalOldFormat(t *testing.T) {
 	password := []byte("gopher")
 	for _, k := range testdata.PEMEncryptedKeys {
-		pk, err := ParseEncryptedRawPrivateKey(k.PEMBytes, []byte(k.EncryptionKey))
-		require.NoError(t, err, "error parsing "+k.Name)
-		require.NotNil(t, pk, "nil return from parsing "+k.Name)
+		// ed25519 is only specified in the new format
+		if k.Name == "ed25519-openssh-encrypted-aes256-cbc" || k.Name == "ed25519-openssh-encrypted-aes256-ctr" {
```
Comment (pquerna) [inline marshal_test.go]: maybe we should just have two test arrays?  Or make a format_type, rather than looking for specific names?

Comment (mavimo) [reply marshal_test.go]: I keep the same approach that we have before, but totally agree on split it in two arrays. Do you suggest to refactor it in the same PR o move in a different one?

File: go.mod (hunk 1 above)
Comment (mavimo) [reply go.mod]: Since golang.org/x/crypto do not have tag I'm not sure we can define the min supported version without specify the specific commit hash

Comment (pquerna) [reply go.mod]: How about we skip the go.mod / vendor modules changes for this PR -- keep it focused on the new algo + testing changes to support testing it (i'm fine with the test package rename).

I'll do some poking on how to best support the go.mod going forward, and that should be an easy separate PR I think once I understand the choices.

**Comment by pquerna** [comment]
went to look this up, openssh changed the default cipher in version 7.6:
https://github.com/openssh/openssh-portable/commit/0f3455356bc284d7c6f4d3c1614d31161bd5dcc2

**Comment by mavimo** [comment]
@pquerna the issue seems related to a security issue that affect CBC algo that can be easily cracked:

 - https://www.openssh.com/txt/cbc.adv

> Ciphers aes128-ctr,aes256-ctr,arcfour256,arcfour,aes128-cbc,aes256-cbc
> ...
> AES CTR mode and arcfour ciphers are not vulnerable to this attack at
all. These may be preferentially selected by placing the following
directive in sshd_config and ssh_config:

NB: I'm not a crypto expert, so maybe I'm wrong 😱

File: go.mod (hunk 1 above)
Comment (mavimo) [reply go.mod]: Ok, I'll revert changes related to go-mod and travis CI

**Comment by mavimo** [comment]
@pquerna done, I need to remove support for golang < 1.9.x since the `golang.org/x/crypto` package need `math/bits` that was included in golang 1.9:

 - https://github.com/golang/go/issues/18616

Last item in CR is the test scenario split.

I will prefer move it in a different PR (where I consider to introduce integration test that allow openssh to create some random key during CI and test that process it), WDYT?

**Comment by pquerna** [comment]
Thank you for the PR!

**Pull request merged**